designs/nodeclaim-limits.md

```diff
@@ -0,0 +1,54 @@
+# NodeClaim Limits
+
+**NodeClaim**: Karpenter’s representation of a Node request. A NodeClaim declaratively describes how the node should look, including instance type requirements that constrain the options that the Cloud Provider can launch with. For the AWS Cloud Provider, the instance requirements of the NodeClaim translate to the requirements of the CreateFleet request. More details can be found [here](https://github.com/aws/karpenter-provider-aws/blob/main/designs/node-ownership.md).
+
+## Motivation
+
+The Karpenter NodePool Spec allows for specifying [provisioning limits](https://karpenter.sh/docs/concepts/nodepools/#speclimits) which can be used to limit based on any resource type provided by the cloudprovider.  If a limit is exceeded, node provisioning is prevented until more resources become available (e.g. node terminations).
+
+This works for users who are purely concerned about limiting resource usage within a NodePool, but doesn't work for specifying a global limit across all NodePools or for limiting a higher level representation of usage such as the number of nodes in a cluster (without also reducing flexibility of instance types).
+
+## Use-Cases
+1. I'm using a shared network with limited IP pools and need to ensure Karpenter's usage doesn't consume too many IPs. Each Node consumes a minimum number of IPs but low-level resource usage for things like CPU/Memory differs between instance types.
+2. I have a special NodePool (for example a set of high performance nodes or AWS Reserved Instances) with a finite amount of compute available. I need a way to prevent Karpenter from launching too many nodes of this type. The only to do this is by calculating against an avialable resource like number of CPU's and adding this as a limit. This doesn't target the actual limiter which is number of instances rather than resource consumption.
+3. My CNI can't support more than 400 nodes but I'm not concerned about resource usage. I have no way to restrict Karpenter to this limit without severly limiting instance type options.
```

Comment:
A bit more on the history of this limitation. 

With kubebnet, only the nodes recieve an IP in the subnet. Pods can't communicate directly with each other and instead rely on UDR and IP forwarding handle connectivity between pods across nodesl. 


Azure supports a maximum of 400 routes in a UDR, so you can't have an AKS cluster larger than 400 nodes if you are using kubenet. 

Thats where this limitation comes from